website/addons/base/testing/models.py

```diff
@@ -390,3 +394,193 @@ def test_after_delete(self):
+
+        # instance was updated
+        assert_equal(
+            self.node_settings.list_id,
+            'fake-folder-id',
+        )
+
+        # user_settings was updated
+        # TODO: the call to grant_oauth_access should be mocked
+        assert_true(
+            self.user_settings.verify_oauth_access(
+                node=self.node,
+                external_account=self.external_account,
+                metadata={'folder': 'fake-folder-id'}
+            )
+        )
+
+        log = self.node.logs[-1]
+        assert_equal(log.action, '{}_folder_selected'.format(self.short_name))
+        assert_equal(log.params['folder_id'], folder_id)
+        assert_equal(log.params['folder_name'], folder_name)
+
+
+class CitationAddonProviderTestSuiteMixin(OAuthCitationsTestSuiteMixinBase):
+
+    @abc.abstractproperty
+    def ApiExceptionClass(self):
+        pass
+
+    def setUp(self):
+        super(CitationAddonProviderTestSuiteMixin, self).setUp()
+        self.provider = self.OAuthProviderClass()
+
+    @abc.abstractmethod
+    def test_handle_callback(self):
+        pass
+
+    def test_citation_lists(self):
+        mock_client = mock.Mock()
+        mock_folders = [MockFolder()]
+        mock_list = mock.Mock()
+        mock_list.items = mock_folders
+        mock_client.folders.list.return_value = mock_list
+        mock_client.collections.return_value = mock_folders
+        self.provider._client = mock_client
+        mock_account = mock.Mock()
+        self.provider.account = mock_account
+        res = self.provider.citation_lists(self.ProviderClass()._extract_folder)
+        assert_equal(res[1]['name'], mock_folders[0].name)
+        assert_equal(res[1]['id'], mock_folders[0].json['id'])
+
+    def test_client_not_cached(self):
+        # The first call to .client returns a new client
+        with mock.patch.object(self.OAuthProviderClass, '_get_client') as mock_get_client:
+            mock_account = mock.Mock()
+            mock_account.expires_at = datetime.datetime.now()
+            self.provider.account = mock_account
+            self.provider.client
+            mock_get_client.assert_called
+            assert_true(mock_get_client.called)
+
+    def test_client_cached(self):
+        # Repeated calls to .client returns the same client
+        with mock.patch.object(self.OAuthProviderClass, '_get_client') as mock_get_client:
+            self.provider._client = mock.Mock()
+            res = self.provider.client
+            assert_equal(res, self.provider._client)
+            assert_false(mock_get_client.called)
+
+    def test_has_access(self):
+        mock_client = mock.Mock()
+        mock_client.folders.list.return_value = self.ApiExceptionClass({'status_code': 403, 'text': 'Mocked 403 ApiException'})
+        mock_client.collections.return_value = self.ApiExceptionClass({'status_code': 403, 'text': 'Mocked 403 ApiException'})
+        self.provider._client = mock_client
+        self.provider._client
```

Comment:
Should this be accessing `.client` to invoke the error? Also, I can't get this test to fail; the assert_raises seems to be a no-op.  Even if I comment out all the setup code, it passes.